### PR TITLE
parse data for interactive map

### PR DESCRIPTION
**Pull request by alisa911**
This pull request is related to other requests and should be merged together with them.
https://github.com/osmandapp/web/pull/330
https://github.com/osmandapp/OsmAnd/pull/19088
https://github.com/osmandapp/OsmAnd-tools/pull/756

### CODE REVIEW

**Comment by alisa911** [comment]
https://github.com/osmandapp/OsmAnd-Issues/issues/2355

File: native/src/commonOsmAndCore.h [hunk 1]
```diff
@@ -352,6 +352,44 @@ class MapDataObject {
...
+	int iconSize;
+
+	// text
+	std::string text;
+	int textSize;
+	int textOnPath;
+	
+	RenderableObject(MapDataObject* _obj) : obj(_obj) {
+		if (obj) {
+			id = obj->id;
```
Comment (ivanPyrohivskyi) [inline native/src/commonOsmAndCore.h:382]: Why need to copy points, types and additionalTypes if you can just create methods getPoints(), getTypes() etc. ?

File: native/src/rendering.cpp [hunk 2]
```diff
@@ -1353,32 +1428,33 @@ void doRendering(std::vector<FoundMapDataObject>& mapDataObjects, SkCanvas* canv
...
-	drawIconsOverCanvas(rc, req, canvas);
+	drawIconsOverCanvas(rc, req, canvas, renderableObjects);
 
 	rc->textRendering.Start();
-	drawTextOverCanvas(rc, req, canvas);
+	drawTextOverCanvas(rc, req, canvas, renderableObjects);
 	rc->textRendering.Pause();
 
+	updateTextTile(renderableObjects, rc);
+
```
Comment (ivanPyrohivskyi) [inline native/src/rendering.cpp:1457]: Free memory from all RenderableObject*

File: native/src/textdraw.cpp [hunk 3]
```diff
@@ -1,22 +1,23 @@
...
 #include <SkTextBlob.h>
+#include <SkTypeface.h>
+#include <SkTypes.h>
+#include <SkUtils.h>
 #include <hb-ot.h>
 #include <math.h>
 #include <time.h>
 
 #include <algorithm>
+#include <iostream>
```
Comment (ivanPyrohivskyi) [inline native/src/textdraw.cpp:20]: Redundant

File: native/src/common.cpp [hunk 4]
```diff
@@ -1,6 +1,7 @@
 #include "CommonCollections.h"
 #include "Logging.h"
 #include "commonOsmAndCore.h"
+#include "json.hpp"
```
Comment (ivanPyrohivskyi) [inline native/src/common.cpp:4]: You include new hpp file,
Will be exception during linking for iOS (core/CMakeLists.txt)

File: native/src/rendering.cpp [hunk 5]
```diff
@@ -606,17 +609,24 @@ int assignOnewayColor(MapDataObject* mObj, RenderingRuleSearchRequest* req, Rend
...
 void drawPolyline(MapDataObject* mObj, RenderingRuleSearchRequest* req, SkCanvas* cv, SkPaint* paint,
-				  RenderingContext* rc, tag_value pair, int layer, int drawOnlyShadow) {
+				  RenderingContext* rc, tag_value pair, int layer, int drawOnlyShadow,
+				  std::unordered_map<int64_t, std::unique_ptr<RenderableObject>>& renderableObjects) {
 	size_t length = mObj->points.size();
 	if (length < 2) {
 		return;
 	}
+
+	RenderableObject* rObj = new RenderableObject(mObj);
```
Comment (ivanPyrohivskyi) [inline native/src/rendering.cpp:619]: Better to change on unique_ptr

File: native/src/rendering.cpp [hunk 6]
```diff
@@ -948,27 +991,29 @@ void drawPoint(MapDataObject* mObj, RenderingRuleSearchRequest* req, SkCanvas* c
 		ico->intersectionSizeFactor = req->getFloatPropertyValue(req->props()->R_INTERSECTION_SIZE_FACTOR, 1);
 		ico->intersectionMargin = getDensityValue(rc, req, req->props()->R_INTERSECTION_MARGIN);
 		ico->secondOrder = ((mObj->id % 10000) << 8) + typeInd;
-		if (ico->order >= 0) rc->iconsToDraw.push_back(ico);
+		if (ico->order >= 0) {
+			rc->iconsToDraw.push_back(ico);
+			renderableObjects.emplace(ico->object.id, rObj);
```
Comment (ivanPyrohivskyi) [inline native/src/rendering.cpp:996]: std::move(

File: native/src/common.cpp (hunk 4 above)
Comment (alisa911) [reply native/src/common.cpp:4]: Max didn't reproduce possible problems

**Review by ivanPyrohivskyi — APPROVED** [review]

File: native/src/rendering.cpp [hunk 7]
```diff
@@ -775,14 +786,19 @@ bool contains(vector<pair<int, int>>& points, int x, int y) {
...
 
 void drawPolygon(MapDataObject* mObj, RenderingRuleSearchRequest* req, SkCanvas* cv, SkPaint* paint,
-				 RenderingContext* rc, tag_value pair, const MapDataObjectPrimitive& prim) {
+				 RenderingContext* rc, tag_value pair, const MapDataObjectPrimitive& prim,
+				 std::unordered_map<int64_t, std::unique_ptr<RenderableObject>>& renderableObjects) {
 	size_t length = mObj->points.size();
 	if (length <= 2) {
 		return;
 	}
+	RenderableObject* rObj = new RenderableObject(mObj);
```
Comment (vshcherb) [inline native/src/rendering.cpp:795]: RenderableObjects always created even if they are not used for Android devices

File: native/src/common.cpp (hunk 4 above)
Comment (vshcherb) [reply native/src/common.cpp:4]: Hpp is a header file looks like so it's not necessary to add

File: native/src/rendering.cpp [hunk 8]
```diff
@@ -410,6 +411,7 @@ void renderText(MapDataObject* obj, RenderingRuleSearchRequest* req, RenderingCo
 							fillTextProperties(rc, dupInfo, req, x, y);
 							dupInfo->secondOrder = ((obj->id % 10000) << 8) + k;
 							rc->textToDraw.push_back(dupInfo);
+							renderableObjects.emplace(dupInfo->object.id, std::move(rObj));
```
Comment (vshcherb) [inline native/src/rendering.cpp:414]: All renderableObjects.emplace should have a flag to add or not to add ? (to not slow down in case it's not needed)

**Pull request merged**